man/foreign.doc

```diff
@@ -1906,33 +1907,40 @@ PL_unify_functor().
     \cfunction{int}{PL_unify_list}{term_t ?l, term_t -h, term_t -t}
 Unify \arg{l} with a list-cell ({\tt ./2}). If successful, write a
 reference to the head of the list into \arg{h} and a reference
-to the tail of the list into \arg{t}. This reference may be used for
+to the tail of the list into \arg{t}. This reference to \arg{h} may be used for
 subsequent calls to this function. Suppose we want to return a list of
 atoms from a \ctype{char **}. We could use the example described by
-PL_put_list(), followed by a call to PL_unify(), or we can use the code
+PL_cons_list(), followed by a call to PL_unify(), or we can use the code
 below. If the predicate argument is unbound, the difference is minimal
-(the code based on PL_put_list() is probably slightly faster). If the
+(the code based on PL_cons_list() is probably slightly faster). If the
 argument is bound, the code below may fail before reaching the end of
 the word list, but even if the unification succeeds, this code avoids a
 duplicate (garbage) list and a deep unification.
 
 \begin{code}
 foreign_t
 pl_get_environ(term_t env)
-{ term_t l = PL_copy_term_ref(env);
```

Comment:
I'm curious why this (rather old) code has PL_copy_term_ref() ... is it because when you wrote the original code, you were going to use PL_cons_list() on `l` or is there something more subtle, where the PL_copy_term() is actually needed? (Code examples in ffi_test.c -- ffi_get_environ3_() and ffi_write_atoms_(), which correspond to examples  put_list() and pl_write_atoms() in foreign.doc.)

Reply:
Oops.   I should have read more carefully.   The PL_copy_term_ref() is needed as you are not allowed to do anything but unify against the argument term references.   PL_unify_list() unifies, but also overwrites where the term reference points at through the 3rd argument, so this is not allowed.   What actually happens is that you destructively overwrite the predicate argument.

I'd have to think about a use case where this actually breaks things.   That might not exist in this case because the argument reference of the foreign predicate is a reference pointer to the argument used in the parent environment, so after the first cell, the one in the parent environment still points to the whole list and the one in the foreign predicate points at the tail.   It probably only upsets stack inspection in this case.   Just for purity, it shouldn't be used.

Note that the `term_t` you get as argument is a direct reference into the environment of the predicate.   That is unlike `term_t` created using `PL_new_term_ref()`, etc, which are term references that are allocated outside the normal Prolog predicate environments and these may thus safely be overwritten and reused.

I guess you want to update both the code and document this?

Reply:
Yes, I'll update both the code and document this. (Hopefully, later today)